polar-core/src/numerics.rs

```diff
@@ -61,9 +66,67 @@ impl PartialEq for Numeric {
     }
 }
 
+impl Eq for Numeric {}
+
 /// There are 53 bits of mantissa in an IEEE 754 double precision float.
 const MOST_POSITIVE_EXACT_FLOAT: i64 = 1 << 53;
 
+/// -i64::MIN is 2**63. The maximum positive i64 is 2**63 - 1, but this
+/// isn't representable as a double. So, we first cast i64::MIN to f64
+/// then flip the sign to get 2 ** 63.
+const MOST_POSITIVE_I64_FLOAT: f64 = -(i64::MIN as f64);
+const MOST_NEGATIVE_I64_FLOAT: f64 = i64::MIN as f64;
+
+impl Hash for Numeric {
+    fn hash<H>(&self, state: &mut H)
+    where
+        H: Hasher,
+    {
+        match self {
+            Numeric::Integer(i) => {
+                discriminant(self).hash(state);
+                *i as u64
+            }
+            Numeric::Float(f) => match f.classify() {
+                FpCategory::Zero => {
+                    // Canonicalize zero representations.
+                    discriminant(&Numeric::Integer(0)).hash(state);
+                    0u64
+                }
+                FpCategory::Nan => {
+                    // Randomize NaN hashes so they always miss.
+                    discriminant(&FpCategory::Nan).hash(state);
+                    f64::from_bits(f.to_bits() | random::<u64>()).to_bits()
```

Comment:
Not sure what ORing with f.bits() is achieving (`f64::NAN.to_bits()` returns `111111111111000000000000000000000000000000000000000000000000000`), so it's just setting the top 12 bits to 1?)

Reply:
No, in a `NaN` the exponent bits (11 of the top 12) are always 1. So this keeps it a `NaN`, but randomizes its payload.

Reply:
Shoot. Realized just now that if the `NaN` is already all 1s, this will not do anything. Need to `xor` it.